[CAMEL-15804] - DataSonnet Expression Language Support

jam01 opened the pull request:
Hey guys, as we had mentioned at some point, we wanted to see about adding support for DataSonnet as a camel expression language. To that effect we've implemented a camel-datasonnet component.

For the most part we feel it's about ready to go. There may be some documentation that needs to be revised, or we might decide to rename a method, but there won't be major changes from what we have now. Our idea was to get the work in front of you guys early in case our implementation approach was incorrect or something has changed from when we first started, considering the work that's been ongoing as part as the last release or two. We already saw some changes in the reifier, for example, when rebasing to 3.7-snapshot.

One thing that we should note is that currently the maven datasonnet dependency points to a version off our maven repository. This is temporary while finalize our v2 release into maven central, shortly. The version currently used should still be publicly available though if you wanted ro run the code locally.

Let us know if you have any questions or feedback on this. Otherwise we hope to mark the PR as ready relatively soon.

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <additionalClasspathElements>
+                        <additionalClasspathElement>${project.basedir}/src/test/resources/dslibs.jar</additionalClasspathElement>
```
What is this library here?

jam01 replied:
It's there to test that we can find supporting library files even within a jar. I'm not sure how much of an use case this is to be honest, if it's not good with you guys we can remove it...

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>org.scala-lang</groupId>
+            <artifactId>scala-library</artifactId>
+            <version>2.13.3</version>
```
If we need this, we need to move the property version placeholder in parent/pom

jam01 replied:
Oh I see... I think that might break some things. I ran into this PR some time ago https://github.com/apache/camel/pull/2060

oscerd replied:
You can also specify a single property placeholder in parent/pom only for camel-datasonnet

jam01 replied:
OK so it seems camel-flink is the only other component using scala-library, unfortunately it depends on an earlier version than what datasonnet-mapper requires. I bumped the parent/pom version to the latest that camel-flink can use (2.12.12) and added a scala-datasonnet-version placeholder in parent/pom that's a few later (2.13.3). If the camel-flink dependency catches up to 2.13 then we could then make a single placeholder. Let me know if this is ok.

jam01 replied:
BTW the commit is https://github.com/apache/camel/pull/4561/commits/704e851cfa7846ce33a5ad6da4d594128c86446f and I ran `mvn clean install -P fastinstall` at the root and everything checks out green

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest</artifactId>
+            <version>2.2</version>
```
There is a placeholder in parent/pom

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>com.datasonnet</groupId>
+            <artifactId>datasonnet-mapper</artifactId>
+            <version>0.6.0-MS3</version>
```
It should go as a placeholder in parent/pom

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>2.4</version>
```
ditto

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>io.github.classgraph</groupId>
+            <artifactId>classgraph</artifactId>
+            <version>4.8.52</version>
```
ditto

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+        <dependency>
+            <groupId>org.skyscreamer</groupId>
+            <artifactId>jsonassert</artifactId>
+            <version>1.5.0</version>
```
ditto

oscerd commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,142 @@
+
+    <repositories>
+        <repository>
+            <id>ms3-nexus</id>
```
We essentially avoid external repository for multiple reasons: our experience with 3rd parties repository was really bad, so if possible please release in maven central.

jam01 replied:
Completely understood. We should be releasing the version we need in the next few days.

jam01 replied:
released :)

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/Datasonnet.java:
```diff
@@ -0,0 +1,20 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/Datasonnet.java:
```diff
@@ -0,0 +1,20 @@
+import org.apache.camel.support.language.LanguageAnnotation;
+
+/**
+ * Datasonnet TODO: 7/21/20 docs
```
remove TODO

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetConstants.java:
```diff
@@ -0,0 +1,6 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetExpression.java:
```diff
@@ -0,0 +1,240 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetExpression.java:
```diff
@@ -0,0 +1,240 @@
+            Objects.requireNonNull(expression, "String expression property must be set!");
+
+            Document<?> result = doEvaluate(exchange);
+            // TODO: 9/8/20 see if we can offload some of this to a Document specific custom type converter
```
Remove TODO

oscerd commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetLanguage.java:
```diff
@@ -0,0 +1,130 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/main/scala/org/apache/camel/language/datasonnet/CML.scala:
```diff
@@ -0,0 +1,58 @@
+package org.apache.camel.language.datasonnet
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/CamelDatasonnetTest.java:
```diff
@@ -0,0 +1,141 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/ExpressionsInJavaTest.java:
```diff
@@ -0,0 +1,64 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/Gizmo.java:
```diff
@@ -0,0 +1,94 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/Manufacturer.java:
```diff
@@ -0,0 +1,48 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/PropertiesTest.java:
```diff
@@ -0,0 +1,36 @@
+package org.apache.camel.language.datasonnet;
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/javaTest.json:
```diff
@@ -0,0 +1,15 @@
+{
```
I guess a license should be good here to

oscerd commented on components/camel-datasonnet/src/test/resources/namedImports.ds:
```diff
@@ -0,0 +1,11 @@
+local testlib1 = import 'testlib.libsonnet';
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/namedImportsFS.ds:
```diff
@@ -0,0 +1,7 @@
+local testlib3 = import 'testlib3.libsonnet';
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/namedImports_result.json:
```diff
@@ -0,0 +1 @@
+{"Lib1JAR":"Hello, World","Lib2JAR":"Bye, World","Lib3FS":"Hello, World : TestLib3","Lib4FS":"Bye, World : TestLib4"}
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/org/apache/camel/language.datasonnet/camel-context.xml:
```diff
@@ -0,0 +1,155 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/payload.csv:
```diff
@@ -0,0 +1,2 @@
+account,firstName,lastName
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/payload.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/readCSVTest.ds:
```diff
@@ -0,0 +1,3 @@
+{
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/readJavaTest.ds:
```diff
@@ -0,0 +1,12 @@
+/** DataSonnet
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/readXMLExtTest.ds:
```diff
@@ -0,0 +1,6 @@
+/** DataSonnet
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/readXMLExtTest.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/simpleMapping.ds:
```diff
@@ -0,0 +1,8 @@
+{
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/simpleMapping_payload.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/simpleMapping_result.json:
```diff
@@ -0,0 +1 @@
+{"count":1,"isActive":true,"testVar":"HelloWorld","uid":123,"uname":"JavaDuke","fullName": "DataSonnet"}
```
Add ASF license please

jam01 replied:
Oh is there a way to add license to JSON files?

oscerd replied:
Not sure if it is feasible, probably not :-)

oscerd commented on components/camel-datasonnet/src/test/resources/testlib3.libsonnet:
```diff
@@ -0,0 +1,4 @@
+{
```
Add ASF license please

oscerd commented on components/camel-datasonnet/src/test/resources/writeJavaTest.ds:
```diff
@@ -0,0 +1,12 @@
+/** DataSonnet
```
Add ASF license please

oscerd commented on core/camel-core-model/src/main/java/org/apache/camel/model/language/DatasonnetExpression.java:
```diff
@@ -0,0 +1,95 @@
+package org.apache.camel.model.language;
```
Add ASF license please

oscerd commented on parent/pom.xml:
```diff
@@ -2642,6 +2647,11 @@
+			<dependency>
+        <groupId>org.apache.camel</groupId>
+        <artifactId>camel-datasonnet</artifactId>
+        <version>${project.version}</version>
```
No need for the dep here, it has already been added on line 1255

oscerd reviewed (changes requested):
Some findings and question. Lets wait for others too. @davsclaus can you have a look?

oscerd commented:
@davsclaus can you have a look too?

jam01 commented:
Hey guys, we released to maven central the latest version of the datasonnet-mapper and updated the property on here. We also marked as resolved the comments related to ASF license on JSON or CSV files since I don't think that's feasible, everything else I think we've addressed.

I've not included any of the generated files as to avoid conflicts with changes from other ongoing work on the main branch. If you'd like me to include that just let me know. Otherwise I'll get one last thumbs up from the team before marking the PR as ready :)

oscerd commented:
I think we should be fine. I'll have to create a Spring-boot starter on camel-spring-boot repository for this, but we can think about it later.

davsclaus commented:
Just a quick glimpse (this PR is long and many comments already) 

- there are still some 3rd party maven repo in the pom.xml
- scala code (is it possible to port this to java so we can maintain the code - camel is primary java)

davsclaus commented on components/camel-datasonnet/src/main/docs/datasonnet-language.adoc:
```diff
@@ -0,0 +1,193 @@
+<route>
+    <from uri="queue:foo"/>
+    <filter>
+        <datasonnet>ds.arrays.firstWith(body.lineItems, function(item) item > 100) != null</datasonnet>
```
In XML I think > needs to be escaped

JakeMHughes replied:
I dont think it does ( unless this ones doing something special that im not aware of) 
The [soap](https://github.com/apache/camel/tree/master/components/camel-soap/src/main/docs/soapjaxb-dataformat.adoc) component docs has some XML in it which is not escaped
And if you meant the `>` inside the value then it looks like [groovy](https://github.com/apache/camel/tree/master/components/camel-groovy/src/main/docs/groovy-language.adoc) docs has an example of that

davsclaus replied:
Ah yeah there may be some mistakes in other parts of the docs. However you need to escape this in XML, but then they are harder to read, > vs &gt; - but we can leave this as-is.

How is it going with all the other review comments?

davsclaus commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,134 @@
+
+    <repositories>
+        <repository>
+            <id>ms3-nexus</id>
```
I assume this 3rd party maven repo can be removed

davsclaus commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,134 @@
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>net.alchim31.maven</groupId>
```
Can we avoid having scala source code. We want everyone to be able to maintain the code and for that we prefer Java

jam01 replied:
addressed :)

davsclaus commented on components/camel-datasonnet/pom.xml:
```diff
@@ -0,0 +1,134 @@
+    <name>Camel :: DataSonnet</name>
+    <description>Camel DataSonnet support</description>
+
+    <properties>
```
Remove these as this is configured in some parent/root pom file

jam01 replied:
Hey Claus, by 'these' you mean the maven compiler properties, correct? The way github shows comments can be confusing sometimes...

davsclaus replied:
Yes its the `<properties>` that can be removed

davsclaus commented on components/camel-datasonnet/src/main/docs/datasonnet-language.adoc:
```diff
@@ -0,0 +1,193 @@
+    <from uri="queue:foo"/>
+    <filter>
+        <datasonnet bodyMediaType="application/xml" outputMediaType="application/json" resultTypeName="java.lang.String" >
+            ds.filter(body.lineItems, function(item) item > 100)
```
Same as before

jam01 replied:
marking as resolved for legibility as stated previously

davsclaus commented on components/camel-datasonnet/src/main/docs/datasonnet-language.adoc:
```diff
@@ -0,0 +1,193 @@
+datasonnet("body.foo", String.class)
+----
+
+In Spring DSL you use the *resultType* attribute to provide a fully
```
Spring DSL -> XML DSL (that is the term we favour)

davsclaus commented on components/camel-datasonnet/src/main/docs/datasonnet-language.adoc:
```diff
@@ -0,0 +1,193 @@
+|cml.exchangeProperty |key for property |String |Will return the exchange property.
+|===
+
+Here's an example showing some of these functions in use:
```
I wonder if we can make a better example as using datasonnet to set the body via a property is not something you would do. I assume you can use these functions in those previous example. For example the value of 100 can be a property

jam01 replied:
changed to `.setBody(datasonnet("'hello, ' + cml.properties('toGreet')", String.class))` let us know if that works

davsclaus commented on components/camel-datasonnet/src/main/docs/datasonnet-language.adoc:
```diff
@@ -0,0 +1,193 @@
+If you use Maven you could just add the following to your `pom.xml`,
+substituting the version number for the latest and greatest release (see
+the download page for the latest versions).
+
```
Very good documentation - thank you

davsclaus commented on components/camel-datasonnet/src/main/java/org/apache/camel/language/datasonnet/DatasonnetExpression.java:
```diff
@@ -0,0 +1,255 @@
+    private static final Map<String, String> CLASSPATH_IMPORTS = new HashMap<>();
+
+    static {
+        LOGGER.debug("One time classpath search...");
```
I would like to see this moved to the Language class, and do this in its doInit method.

jam01 replied:
moved :)

davsclaus commented on components/camel-datasonnet/src/main/scala/org/apache/camel/language/datasonnet/CML.scala:
```diff
@@ -0,0 +1,74 @@
+import sjsonnet.Std.builtin
+import sjsonnet.{Materializer, Val}
+
+object CML extends Library {
```
Port this to plain Java code so everyone can maintain it

davsclaus commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/CamelDatasonnetTest.java:
```diff
@@ -0,0 +1,157 @@
+                "direct:namedImports");
+    }
+
+    // TODO: 9/8/20 need to pass result type param to language
```
Fix this TODO

davsclaus commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/CamelDatasonnetTest.java:
```diff
@@ -0,0 +1,157 @@
+        mock = getMockEndpoint("mock:direct:end");
+        Exchange exchange = mock.assertExchangeReceived(mock.getReceivedCounter() - 1);
+        String response = exchange.getIn().getBody().toString();
+        System.out.println("RESPONSE IS " + response);
```
Remove system out or do a LOG

davsclaus reviewed (changes requested):
See comments

davsclaus commented:
Any update on the other review comments?

JakeMHughes commented:
I think we've got most of the smaller ones done.
Still working on the Scala -> Java conversion though.

davsclaus commented:
Any update on when the last bits are complete ?

JakeMHughes commented:
Were pretty much wrapping it up and I think were looking to be finished sometime today / tomorrow

jam01 commented:
@davsclaus we've migrated the Scala code to Java. It brought up a couple of other changes we decided we decided to address at once. We're merging things up now and plan to have the new mapper version released and reflected here early tomorrow.

fugu13 commented:
Hi, I'm one of the creators of DataSonnet, and I wanted to come by and leave a comment on how excited I am to possibly get it into Camel. Being able to apply the very nicely designed JSonnet language to data transformation problems has been extremely powerful in work my coworkers do, and we've already seen adoption by quite a few other folks working in similar spaces. We're already using it in Camel ourselves with an internal implementation, and it's been a great experience.

We worked with @jam01 and @JakeMHughes to add a bunch of functionality to DataSonnet in anticipation for this PR they've put together, including an adapter that got rid of the need to use Scala for defining DataSonnet libraries, and we're hoping lots of other folks using Camel find it useful too.

davsclaus commented on core/camel-core-model/src/main/java/org/apache/camel/builder/DatasonnetBuilder.java:
```diff
@@ -0,0 +1,56 @@
+import org.apache.camel.Expression;
+import org.apache.camel.model.language.DatasonnetExpression;
+
+public class DatasonnetBuilder extends ValueBuilder {
```
I dont see any extra value of having its own Builder class, we do not really do this for other data formats - only xml and xpath have a special builder etc. Can this be removed

jam01 replied:
I think we based this on one of those builders. The reason comes from bodyMediaType and outputMediaType being both individually optional, so we could have any of these:

```
// here we don't use the body, and the default output is fine
.transform(datasonnet("hello, world", String.class))
```
```
// here we need to be explicit about the input, maybe the content-type header does not apply, for example
.transform(datasonnet("payload", String.class).bodyMediaType("application/json")) 
```
```
// here we want the component to look in other locations for the input (eg. content-type header)
// but we need a different output than default
.transform(datasonnet("payload", String.class).outputMediaType("application/xml")))
```
```
// here both are needed
.transform(datasonnet("some-expression", String.class)
    .outputMediaType("application/json")
    .bodyMediaType("application/xml")))
```

So the value is being able to be selectively explicit with the input and output types. Otherwise we'd have to do 

```
// we could have a 'all defaults' type of constructor
.transform(datasonnet("hello, world", String.class))
```
```
.transform(datasonnet("payload", String.class, "application/json", null)) 
```
```
.transform(datasonnet("payload", String.class, null, "application/json")) 
```
```
.transform(datasonnet("some-expression", String.class, "application/json", "application/xml"))
```

So I guess ultimately it's about readability since both options are functionally the same. I think a predicate is one particular scenario that might be confusing

```
// if we needed to be explicit on the body, do we leave the output as null so the default is used?
// or do we make it Java as it should be, either way the component forces Java for predicates
// but the extra null could be confusing to devs, or misleading if they wrote something other than null or Java
.when(datasonnet("payload.predicate == 'yes'", "application/json", null))
   ...
// the output type would be ignored
.when(datasonnet("payload.predicate == 'yes'", "application/json", "application/json"))
```

I hope that expresses the value we see in the Builder. 

This wasn't actually a widely discussed decision internally, I just saw those other Builders you mentioned and seemed like the right way to achieve what I described. So we're all ears :)

If there's a more "camel way" to achieve this we'd be happy to refactor. Or if the Builder class should be moved to the datasonnet component (I think the XML one is) in order to move maintenance there, we can do that. Or if ultimately this doesn't fit with the Builder's design we can remove it.

davsclaus replied:
Well I am afraid thats how the other data formats do it with optional parameters, is that they have a number of methods for different combinations, and that you would have to pass in null for some of them.

There is no room for having specialized builders at this moment. The architecture of dataformats and how its too tightly coupled into the model/core is something we want to address in future Camel - such as Camel 4 or sooner if we can make it possible in 3.x.

jam01 replied:
Ah! Copy that. Would it be OK if we alternatively move the extra methods to the camel-datasonnet DatasonnetExpression? I see XPath has support for building expressions from both camel-core-model [BuilderSupport](https://github.com/apache/camel/blob/master/core/camel-core-model/src/main/java/org/apache/camel/builder/BuilderSupport.java#L197) and also from camel-xpath's [XPathBuilder](https://github.com/apache/camel/blob/master/components/camel-xpath/src/main/java/org/apache/camel/language/xpath/XPathBuilder.java#L156). The camel-xpath one actually has optional parameter support, just as we were looking for, as showcased in [this test](https://github.com/apache/camel/blob/master/core/camel-core/src/test/java/org/apache/camel/component/dataset/CustomDataSetTest.java#L34). Here's [an example](https://github.com/apache/camel/blob/master/components/camel-saxon/src/test/java/org/apache/camel/language/xpath/XPathHeaderEnableSaxonJavaDslTest.java#L67) of the camel-xpath one being used in a route instead of the RouteBuilder one.

So developers could use the BuilderSupport one if they just want default media types, or the camel-datasonnet one if they want to be explicit in either one or both. Additionally this would allow us to user DataSonnet's formal MediaType objects instead of their String representation. I used the String representation in the BuilderSupport as to not have core-model depend on the DataSonnet Mapper. This would translate into some performance gains as the String will not need to be parsed on each expression evaluation.

davsclaus replied:
Yes this is a good idea

davsclaus commented on components/camel-datasonnet/src/test/java/org/apache/camel/language/datasonnet/CamelDatasonnetTest.java:
```diff
@@ -0,0 +1,164 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class CamelDatasonnetTest extends CamelSpringTestSupport {
+    private MockEndpoint mock;// = getMockEndpoint("mock:direct:end");
```
remove the left over code in the comment

davsclaus commented:
Is this ready for final review and merge?

jam01 commented:
@davsclaus yes, it is :)
